File: src/androidTest/java/at/bitfire/dav4android/exception/ServiceUnavailableExceptionTest.java
/*
 * Copyright © 2013 – 2015 Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.dav4android.exception;

import com.squareup.okhttp.Protocol;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.Response;
import com.squareup.okhttp.internal.http.HttpDate;

import junit.framework.TestCase;

import java.util.Calendar;
import java.util.Date;

public class ServiceUnavailableExceptionTest extends TestCase {

    public void testRetryAfter() {
        Response response = new Response.Builder()
                .request(new Request.Builder()
                        .url("http://www.example.com")
                        .get()
                        .build())
                .protocol(Protocol.HTTP_1_1)
                .code(503)
                .build();

        ServiceUnavailableException e = new ServiceUnavailableException(response);
        assertNull(e.retryAfter);

        response = response.newBuilder()
                .header("Retry-After", "120")
                .build();
        e = new ServiceUnavailableException(response);
        assertNotNull(e.retryAfter);
        assertTrue(withinTimeRange(e.retryAfter, 120));

        Calendar cal = Calendar.getInstance();
        cal.add(Calendar.MINUTE, 30);
        response = response.newBuilder()
                .header("Retry-After", HttpDate.format(cal.getTime()))
                .build();
        e = new ServiceUnavailableException(response);
        assertNotNull(e.retryAfter);
        assertTrue(withinTimeRange(e.retryAfter, 30*60));
    }


    private boolean withinTimeRange(Date d, int seconds) {
        final long msCheck = d.getTime(), msShouldBe = new Date().getTime() + seconds*1000;
        // assume max. 5 seconds difference for test running
        return Math.abs(msCheck - msShouldBe) < 5000;
    }

}

File: src/test/kotlin/at/bitfire/dav4jvm/exception/ServiceUnavailableExceptionTest.kt
/*
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package at.bitfire.dav4jvm.exception

import at.bitfire.dav4jvm.HttpUtils
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Response
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Instant

class ServiceUnavailableExceptionTest {

    val response503 = Response.Builder()
        .request(
            Request.Builder()
                .url("http://www.example.com")
                .get()
                .build()
        )
        .protocol(Protocol.HTTP_1_1)
        .code(503).message("Try later")
        .build()

    @Test
    fun testRetryAfter_NoTime() {
        val e = ServiceUnavailableException(response503)
        assertNull(e.retryAfter)
    }

    @Test
    fun testRetryAfter_Seconds() {
        val response = response503.newBuilder()
            .header("Retry-After", "120")
            .build()
        val e = ServiceUnavailableException(response)
        assertNotNull(e.retryAfter)
        assertTrue(withinTimeRange(e.retryAfter!!, 120))
    }

    @Test
    fun testRetryAfter_Date() {
        val after30min = Instant.now().plusSeconds(30 * 60)
        val response = response503.newBuilder()
            .header("Retry-After", HttpUtils.formatDate(after30min))
            .build()
        val e = ServiceUnavailableException(response)
        assertNotNull(e.retryAfter)
        assertTrue(withinTimeRange(e.retryAfter!!, 30 * 60))
    }


    private fun withinTimeRange(d: Instant, seconds: Long) =
        d.isBefore(
            Instant.now()
                .plusSeconds(seconds)
                .plusSeconds(5)     // tolerance for test running
        )

}